[step 4-6] 방탈출 사용자 예약 미션 제출

kkiseug opened the pull request:
## + 1차 리뷰 반영
#### 리뷰를 위한 [Files changed](https://github.com/woowacourse/spring-roomescape-member/pull/281/files/8cf4223235710e4e510377eb401f0208a938a65b..aa539d63b332391eb21e5acf048741e683874e06)

리뷰 반영과 함께 추가적으로 예외 관련하여 구조를 크게 개선해보았습니다!! 커스텀 예외도 적용해보았어요. 해당 부분도 함께 리뷰 부탁드립니다!

+22:29:06) ExceptionHandler 처리를 까먹었네요!!!!!!!!!!!!! ..


## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
[프롤로그](https://prolog.techcourse.co.kr/studylogs/4312)

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
```text
[x] 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
[x] else 예약어를 쓰지 않는다.
[ ] 모든 원시 값과 문자열을 포장한다.
[ ] 일급 컬렉션을 쓴다.
[ ] 한 줄에 점을 하나만 찍는다.
[x] 줄여 쓰지 않는다(축약 금지).
[ ] 모든 엔티티를 작게 유지한다.
[ ] 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
[ ] getter/setter/프로퍼티를 쓰지 않는다.
```

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [x] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->
### [리뷰를 위한 Files changed](https://github.com/woowacourse/spring-roomescape-member/pull/281/files/8cf4223235710e4e510377eb401f0208a938a65b..5ed33748b788639d4431ae38c6e36ff7439d6a26)
step2로 옮기는 과정에서 step1의 커밋까지 끌고와버렸습니다! 위 링크를 통해 step2 커밋만 적용해놓았습니다.

### Admin과 Member의 분리
저는 Admin과 Member를 각자의 객체로 분리하였습니다. 이렇게 설계한 의도는 애초에 Admin과 Member가 서로 역할이 다르다고 생각했기 때문이에요. Role이라는 단어 자체가 역할을 의미하기도 하고, 이를 상태로 갖는다는 건 레벨 1에서 배운 객체지향의 개념과는 멀어진다고 생각합니다. 그래서 저는 토큰에만 Role에 대한 정보를 담았고, interceptor와 resolver에서 해당 토큰에 따라 Admin인지, Member인지를 판단하고 해당 객체에 대한 id를 넘겨주었습니다.

일단 먼저 한 가지 확실히 할 것은 이번 미션에서는 어드민과 유저의 구현이 크게 다르지 않다는 점입니다. 그래서 중복 코드처럼 보일 수 있습니다. 하지만, 제 생각에 추후 어드민이 가지는 고유한 책임이 생긴다면 이를 Member로 퉁쳐 하나로 합치는 것보다 Admin, Member로 나눈 지금의 설계가 좀 더 객체지향이다. 라고 말할 수 있을 것 같아요.

아쉽게도 이러한 설계가 만능은 아닌게 사소한 단점이 조금 있었습니다.

처음에는 제 설계가 완벽한 줄 알았습니다. 근데 어드민으로 로그인을 하고 유저 페이지로 넘어가면 엉뚱한 유저의 아이디로 로그인되는 상황이 발생했어요. 이는 어드민 식별 id와 유저의 식별 id가 같은 타입이기에 발생한 문제였습니다. 그래서, 지금은 유저 페이지에 접속할 때 member resolver에서 토큰 속 role이 멤버가 아니라면 로그인을 하지 못하도록 막고 있습니다.

처음엔 이런 구조가 썩 맘에 들진 않았습니다. 하지만 지금은 좀 다릅니다. 저는 기본적으로 Admin은 본인의 예약을 하지 못한다. 라고 생각을 했기에 어드민이 My Reservation 페이지를 들어간다는 것 자체가 이상하기도 하고 Admin 계정으로 유저 페이지를 사용할 일이 없다고 생각했어요. 오직 Admin의 페이지에서만 의미가 있다고 생각했습니다. 그래서 이런 구조를 그대로 유지했습니다.

<br/>

### JWT 토큰 구성 요소의 이유
JWT 토큰에는 id 값과 role만 넣어주었습니다. 추가적으로 name 필드가 응답에서 사용되기에 이를 토큰의 claim으로 넣을 수 있지 않을까 생각했지만, name은 변하는 값이기에 넣지 않았어요. 예를 들어 유저가 개명을 하는 상황 또는 이름이 아닌 닉네임으로 본다면, 변경하는 과정에서 항상 해당 유저의 token을 새로 발급해줘야 하는 일이 생깁니다. 변경 로직이 응답에 영향을 받는 것이죠.

거기다가 이런 설계라면 응답 명세가 변할 때마다 해당 JWT 토큰도 변화해야 하는 불필요한 수정이 발생할 수 있습니다. 즉, 검증을 위한 토큰이 응답에 너무 직접적인 관여를 하게 된다고 생각했어요. 그래서 최소한의 정보만 유지하도록 구성했습니다. 단점은 id만 있기에 응답을 위해서 findById를 호출해서 쿼리를 한번 거쳐야 된다는 점이 있지만 정확성이 더 높아졌다고 생각합니다.

반대로 role은 변할 수 있는 값으로 볼 수 있는 것 아니냐고 할 수 있지만, 제 설계는 Admin과 Member를 아예 다른 역할로 보고 있기에 둘이 바뀌는 순간은 전혀 생길 수 없습니다! 그렇기에 변하지 않는 값이에요.


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

kkiseug committed:
feat: setup project

kkiseug committed:
docs: 1단계 요구사항 정리

kkiseug committed:
refactor: time 패키지 reservationTime으로 변경

kkiseug committed:
refactor: application 계층 service 계층으로 변경

kkiseug committed:
fix: 예약 시간 추가 시 응답 코드 수정

- 200 -> 201

kkiseug committed:
fix: 예약 추가 시 응답 코드 수정

- 200 -> 201

kkiseug committed:
fix: 예약 시간 삭제 시 응답 코드 수정

- 200 -> 204

kkiseug committed:
fix: 예약 삭제 시 응답 코드 수정

- 200 -> 204

kkiseug committed:
feat: 시간 유효하지 않은 값 예외처리

kkiseug committed:
test: 기존 테스트 추거

kkiseug committed:
feat: 예약자명, 날짜, 시간 검증 및 예외 처리

kkiseug committed:
feat: 예약 시간 삭제 예외 처리

- 특정 시간에 대한 예약이 존재할 때 해당 시간 삭제 시 예외 발생

kkiseug committed:
feat: 지나간 날짜와 시간에 대한 예약 생성 예외 처리

kkiseug committed:
feat: 중복 예약 시 예외 처리

kkiseug committed:
test: 예약 검증 테스트 구현

kkiseug committed:
test: 예약 시간 검증 테스트 구현

kkiseug committed:
test: 예약 관련 예외 테스트 추가

kkiseug committed:
test: 예약 시간 관련 예외 테스트 추가

kkiseug committed:
chore: sql 스키마 수정

kkiseug committed:
feat: 테마 페이지 요청 응답 기능 구현

kkiseug committed:
feat: 테마 도메인 모델 생성

kkiseug committed:
docs: 테마 요구 사항 추가

kkiseug committed:
feat: 테마 레포지토리 생성

- 테마 저장
- 테마 조회
- 테마 삭제

kkiseug committed:
feat: 테마 서비스 계층 추가

- 테마 추가
- 테마 id로 삭제
- 테마 조회

kkiseug committed:
feat: 테마 삭제 예외 추가

kkiseug committed:
feat: 테마 관리 API 기능 구현

kkiseug committed:
feat: 예약 도메인 테마 객체 추가

kkiseug committed:
refactor: repository 인터페이스 추상화

kkiseug committed:
fix: 도메인 변경에 대한 테스트 코드 수정

kkiseug committed:
refactor: Theme dto 패키지 변경

kkiseug committed:
test: 존재하는 예약의 테마 삭제 예외 테스트 구현

kkiseug committed:
test: 테마 api 통합 테스트 추가

kkiseug committed:
docs: 사용자 관련 기능 정리

kkiseug committed:
feat: 사용자 예약 페이지 응답 기능 구현

kkiseug committed:
feat: 사용자 예약 시 예약 가능 시간 조회 API 기능 구현

kkiseug committed:
docs: 인기 테마 요구사항 정리

kkiseug committed:
chore: 초기 데이터 삽입 sql 파일 생성

kkiseug committed:
fix: 랭킹 조회 js 수정

kkiseug committed:
feat: 인기 테마 조회 API 기능 구현

kkiseug committed:
refactor: 썸네일 경로 수정

- 이미지 추가

kkiseug committed:
fix: 요구사항 변경에 따른 테스트 수정

kkiseug committed:
test: 테마와 날짜에 따른 예약 가능 시간 조회 테스트

kkiseug committed:
test: 인기 테마 조회 테스트

kkiseug committed:
fix: 예약 시 테마 조건 추가

kkiseug committed:
refactor: 예외 처리 핸들러 세분화

kkiseug committed:
remove: 쓸모없는 주석 제거

kkiseug committed:
refactor: 서비스 생성 로직 변경

- 생성 후 findById 호출 삭제
- 생성 시 목적 나타내기 위한 팩토리 메서드 적용

kkiseug committed:
refactor: repository 삭제 시 리턴 타입 변경

- int -> boolean

kkiseug committed:
refactor: final 키워드 적용

kkiseug committed:
refactor: 인기 테마 상위 개수 상수화

kkiseug committed:
refactor: 매직넘버 상수화

kkiseug committed:
refactor: 검증 메서드 분리

kkiseug committed:
refactor: 쓸모없는 try ~ catch 문 제거

kkiseug committed:
docs: README 수정 사항 업데이트

kkiseug committed:
refactor: 매핑 URI 명사형으로 변경

- ThemeController: `/themes/popular` -> `/themes/popular-themes`

kkiseug committed:
refactor: 중복되는 URI @RequestMapping 적용

kkiseug committed:
refactor: id 기준으로 객체 비교하도록 equals 재정의

kkiseug committed:
refactor: 예외 처리 클래스 패키지 세분화

kkiseug committed:
refactor: 예외 핸들러 수정

- NullPointerException 예외 처리 추가
- `@ControllerAdvice` -> `@RestControllerAdvice`

kkiseug committed:
refactor: 검증 예외 도메인 계층에 추가

- PopularThemeResponse 필드 순서 변경 및 순서 변경으로 인한 테스트 수정
- 예외 메시지 일부 수정

kkiseug committed:
refactor: dto 패키지 이동

- presentation 계층으로 이동

kkiseug committed:
refactor: JdbcTemplate -> NamedParameterJdbcTemplate로 변경

kkiseug committed:
fix: 프론트 end point 수정

kkiseug committed:
refactor: 존재 여부 쿼리 수정

- COUNT -> EXISTS

kkiseug committed:
refactor: 예외 처리 순서 변경

- 중복 예약에 대해 먼저 검사하도록 변경

kkiseug committed:
refactor: 쓸모없는 import문 제거 및 final 키워드 적용

kkiseug committed:
refactor: URI 상수화

kkiseug committed:
refactor: 예외 메시지 수정

- 시간, 테마에 대한 예약이 존재하는 경우 삭제 불가 메시지 수정

kkiseug committed:
refactor: 과거 예약 검증 로직 변경

- LocalDateTime을 직접 만들지 않고 Reservation 통해 검증하도록 변경

kkiseug committed:
test: 테스트용 schema 생성

kkiseug committed:
refactor: ReservationService 테스트 페이크 객체 의존하지 않도록 변경

kkiseug committed:
fix: 테스트 스키마 수정

- 테스트마다 스키마 생성 시 drop 하지않아 예외 발생 문제 해결

kkiseug committed:
refactor: service 테스트 fake 객체 의존하지 않도록 변경

- fake 객체 삭제

kkiseug committed:
refactor: 쓸모없는 static 제거

kkiseug committed:
refactor: View 반환 컨트롤러 패키지 변경

kkiseug committed:
feat: 멤버 테이블 생성 및 초기 데이터 추가

- member 테이블 스키마를 schema.sql에 추가
- init.sql에 member 초기 데이터 삽입 쿼리 추가

kkiseug committed:
chore: jwt 라이브러리 의존성 추가

- jjwt 라이브러리 의존성 추가

kkiseug committed:
refactor: DateTime 관련 패키지 수정

- util.time 패키지로 이동

kkiseug committed:
feat: 로그인 페이지 기능 구현

- /login GET 요청 시 로그인 페이지 반환

kkiseug committed:
feat: Member 도메인 관련 기능 구현

- Member 도메인 객체 구현
- MemberRepository 인터페이스 및 구현
- Role Enum 정의

kkiseug committed:
feat: JWT 토큰 관련 기능 구현

- 토큰 발행
- 토큰 검증 후 Subject 추출

kkiseug committed:
feat: 잘못된 토큰 추출 시 예외 처리 구현

- SignatureException 발생 시 401 상태코드와 메시지를 반환

kkiseug committed:
feat: 로그인 기능 구현

- 로그인 및 로그인 상태 확인 api 구현

kkiseug committed:
feat: 쿠키 없을 시 예외 처리 기능 구현

kkiseug committed:
feat: 잡지 않은 예외 공통 처리 기능 구현

kkiseug committed:
refactor: Admin, Member 클래스 분리

- Role 삭제
- Admin 및 Member 클래스로 역할 분리 및 책임 명확하게 변경

kkiseug committed:
feat: 사용자 로그인 및 프로필 활성화

- 4단계 TODO 제거

kkiseug committed:
feat: admin 테이블 추가

- admin 테이블 생성

kkiseug committed:
fix: 데이터 스키마 수정

- member 테이블 role 필드 제거
- admin 테이블 데이터 추가

kkiseug committed:
feat: Admin 관련 Repository, Service 구현

- 이메일 또는 ID로 Admin 조회 및 존재 여부 확인 기능 제공

kkiseug committed:
feat: Member 관련 Repository, Service 구현

- 이메일 또는 ID로 Member 조회 및 존재 여부 확인 기능 제공

kkiseug committed:
feat: jwt 토큰 속 role 추출 기능 구현

- 메서드 이름 getMemberId -> getId로 수정
- 불필요한 import 제거

kkiseug committed:
feat: 로그인 시 Admin, Member를 구분하는 로직 추가

- /login/check에서 role을 추출하여 구분하여 검증하도록 변경

kkiseug committed:
chore: favicon 추가

- I Like Diamond.

kkiseug committed:
feat: 어드민 로그인 페이지 및 관련 스크립트 추가

- 어드민 전용 로그인 페이지(admin/login.html) 생성
- 어드민 스크립트 분리
- `LoginPageController`에 어드민 로그인 페이지 매핑
- 어드민 페이지에서 user-scripts.js 대신 admin-scripts.js 사용

kkiseug committed:
feat: 관리자 로그인 기능 추가 및 회원 로그인 로직 분리

- AdminLoginController를 생성하여 관리자 로그인 및 인증 기능 구현
- 기존 LoginController를 MemberLoginController로 리팩터링하여 회원 로그인 로직을 분리

kkiseug committed:
fix: 로그인 경로 수정

- 관리자 로그인 경로를 "/login"에서 "/admin/login"으로 수정

kkiseug committed:
refactor: Admin 관련 클래스 패키지 구조 변경

- member 패키지 -> admin 패키지로 이동

kkiseug committed:
refactor: 관리자 페이지 경로 수정

- 경로 `/admin` 기반으로 수정
- Login 페이지 명확하도록 Admin 추가

kkiseug committed:
feat: Member와 Admin을 인증하는 ArgumentResolver 추가

- @LoginMember, @LoginAdmin 애노테이션 생성

kkiseug committed:
feat: ArgumentResolver 등록

kkiseug committed:
feat: jwt 토큰 속 id를 가지는 DTO 추가

kkiseug committed:
refactor: 컨트롤러의 쿠키 관련 로직 삭제

- HttpServletRequest 및 쿠키 기반 인증 제거
- @LoginAdmin, @LoginMember 어노테이션과 DTO를 활용한 인증 구조 변경

kkiseug committed:
feat: 예약 생성 관련 프론트 코드 수정

- 사용자 이름 입력 필드 제거
- 관리자 예약 스크립트 변경
- 관리자 예약 스크립트 명세 수정

kkiseug committed:
feat: 모든 회원 조회 기능 추가

- 관리자 예약 생성 시 회원 조회

kkiseug committed:
feat: 예약 생성 시 회원 정보 추가

- 예약 생성 메서드에 회원 ID를 파라미터로 추가
- MemberRepository 의존성 주입 및 회원 정보 조회 로직 추가
- 예약 시 회원 이름을 사용하도록 변경

kkiseug committed:
feat: 사용자 예약 생성 시 명세 변경에 따른 수정

- 사용자 정보를 받도록 파라미터 추가
- ReservationRequest 필드 변경

kkiseug committed:
refactor: 로그인 및 예약 관련 패키지 구조 개선

kkiseug committed:
refactor: argument resolver에서 Role에 대한 검증하도록 변경

kkiseug committed:
feat: 토큰 속 role 검증 인터셉터 구현

kkiseug committed:
refactor: 토큰 역할 정보 수정

- 토큰 생성 시 역할 정보를 "USER"에서 "MEMBER"로 수정

kkiseug committed:
refactor: AdminRoleInterceptor로 클래스명 변경

- AuthHandlerInterceptor를 AdminRoleInterceptor로 클래스명 변경

kkiseug committed:
refactor: 조건별 예약 조회 API 수정

kkiseug committed:
refactor: reservation 필드 수정

- name 필드 제거
- Member 필드 추가

kkiseug committed:
feat: 검색 조건 파라미터 dto 추가

kkiseug committed:
refactor: reservation 스키마 수정

- name 컬럼 제거
- member_id 컬럼 추가

kkiseug committed:
refactor: 검색 조건에 대한 repository 메서드 추가

- reservation과 member 연관 관계에 따라 ROW_MAPPER 수정
- member 정적 팩토리 메서드 추가
- memberRepository ROW_MAPPER 수정

kkiseug committed:
refactor: 초기 데이터 sql 수정

kkiseug committed:
fix: 관리자 페이지 예약 정보 출력 수정

- 'Login' 메뉴의 링크를 '/login'에서 '/admin/login'으로 수정
- 예약 정보 출력 시 사용자 이름을 `item.name`에서 `item.member.name`으로 수정하여 데이터 참조 오류 해결

kkiseug committed:
feat: 예약 조건 검색 기능 및 관련 기능 구현

- ReservationResponse에 Member 정보 포함

kkiseug committed:
feat: 어드민 api 접속 차단 기능 구현

- Interceptor를 이용하여 admin이 아니면 차단

kkiseug committed:
feat: 회원 저장 로직 추가 및 JWT 유효 기간 설정

kkiseug committed:
feat: 회원 정보 추가를 통한 예약 로직 개선

- Reservation에서 회원 정보를 포함하도록 테스트 코드 수정
- JdbcMemberRepository를 활용해 회원 데이터를 처리하도록 변경
- AdminIntegrationTest에서 관리자가 아닌 접근에 대해 401 응답하도록 변경

kkiseug committed:
refactor: 관리자 resolver 필요없는 구문 제거

kkiseug committed:
refactor: 로그인 서비스 로직 분리 및 중복 제거

- AdminLoginController와 MemberLoginController의 공통 로직을 LoginService로 분리
- UnauthorizedException 커스텀 예외로 인증 관련 예외 처리

kkiseug committed:
refactor: auth 패키지 구조 변경 및 토큰 로직 분리

- auth 관련 패키지 구조 수정
- TokenExtractor 클래스를 추가하여 토큰 추출 로직 분리
- JwtTokenManager에 토큰 유효성 검증 메서드 추가

kkiseug committed:
refactor: 컨트롤러 네이밍 수정

- AdminController를 AdminPageController로 이름 변경
- Member 페이지 반환 관련 로직 MemberPageController로 분리

kkiseug committed:
moved: MemberResponse 패키지 이동

kkiseug committed:
feat: 인증 관련 커스텀 예외 처리 기능 구현

- ExceptionHandler에서 UnauthorizedException 처리하도록 함

kkiseug committed:
feat: 권한 관련 예외 처리 추가

- ForbiddenException 클래스 추가
- 관리자 및 사용자 권한 검증 메서드 분리
- GlobalExceptionHandler에 ForbiddenException 처리

kkiseug committed:
refactor: JWT 개인키를 외부 설정으로 변경

- application.properties에서 @Value를 통해 주입하도록 변경

kkiseug committed:
refactor: 쿠키에서 토큰 꺼내는 작업 TokenExtractor로 분리

kkiseug committed:
refactor: @Component 어노테이션 제거

- LoginMemberArgumentResolver
- LoginAdminArgumentResolver

kkiseug committed:
feat: 멤버 도메인 속 원시값 포장

- Name, Email, Password 클래스를 추가하여 멤버 속성들을 객체로 추상화
- Member 클래스에서 String 대신 새로운 객체를 사용하도록 변경
- 이름 길이와 이메일 형식을 검증하는 로직 구현

kkiseug committed:
refactor: 관리자 페이지 회원가입 관련 코드 제거

kkiseug committed:
feat: 회원가입 페이지 응답 구현

kkiseug committed:
fix: JWT 토큰 생성 오류 수정

- 토큰 유효 시간 설정 오류 수정
- SecretKey 설정 방식을 @Value 어노테이션 기반으로 수정

kkiseug committed:
fix: SimpleJdbcInsert 설정 정보 추가

kkiseug committed:
feat: 회원 생성 기능 추가

kkiseug committed:
refactor: JwtTokenManager를 인스턴스 기반으로 리팩토링

- JwtTokenManager의 메서드들을 static에서 인스턴스 메서드로 변경
- @Component 어노테이션 추가 및 WebMvcConfig에서 빈 주입

kkiseug committed:
refactor: 테스트용 스키마 수정

kkiseug committed:
feat: 테스트용 데이터 및 설정 파일 추가

- 테스트 데이터 초기화 SQL 파일 추가 (data.sql)
- 테스트 환경용 secret.key 설정 추가 (application.properties)

kkiseug committed:
fix: 토큰 생성 권한 오류 수정

- ADMIN 권한 대신 MEMBER 권한으로 토큰 생성하도록 변경

kkiseug committed:
fix: 비밀번호 비교 로직 수정

- 기존 코드에서 password 필드 접근 방식 오류 수정
- 비밀번호 비교 시 password 객체 내부 값으로 검증하도록 변경

kkiseug committed:
refactor: 통합 테스트 데이터 수정 및 불필요 테스트 제거

kkiseug commented on src/main/java/roomescape/common/configuration/WebMvcConfig.java:
```diff
@@ -0,0 +1,40 @@
+@Configuration
+public class WebMvcConfig implements WebMvcConfigurer {
+
+    private final LoginAdminArgumentResolver loginAdminArgumentResolver;
+    private final LoginMemberArgumentResolver loginMemberArgumentResolver;
+    private final AdminRoleInterceptor adminRoleInterceptor;
```
`@Value` 어노테이션을 이용해 키를 숨기려면 어쩔 수 없이 빈으로 등록해야 하더라고요! 저는 최대한 WebMvcConfig에서 파라미터를 적게 가져가고 싶었는데, 그러지 못해서 아쉽습니다.

soominsohn replied:
ArgumentResolver, Interceptor 모두 Spring환경에서 통합되어 사용되고 있기 때문에, 빈으로 사용되는게 좋을 것 같아요. 
한 Configuration파일에 파라미터가 많아진다면, ArgumentResolver와 Interceptor 설정파일을 분리하는 것도 방법이 될 수 있어요. 제가 질문 의도를 잘 이해한걸까요?

kkiseug replied:
푸념에 가까운 질문이었는데, 답변이 많은 도움이 되었어요!! 역시 앤지한테서 많이 배웁니다.

soominsohn replied:
저도 율무가 코멘트 남겨주시면 다시 생각해보면서 많이 배웁니다~

kkiseug commented on src/main/java/roomescape/auth/login/infrastructure/interceptor/AdminRoleInterceptor.java:
```diff
@@ -0,0 +1,38 @@
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+            throws Exception {
+        String token = TokenExtractor.extract(request);
```
Interceptor와 Resolver에서의 쿠키 -> 토큰 과정이 중복되는 코드 같다고 생각했어요. 하지만, 호출 시점이 다르기에 중복이 아니라고 생각합니다! 앤지는 어떤가요?

soominsohn replied:
네 저도 율무 의견에 동의합니다~. 하는 일은 같지만 호출하는 곳이 다르기 때문에 중복이 아니라고 생각해요. 😄 
굳이 지금 학습할 필요는 없지만(비추합니다!)~ 아주 나중에 이 중복을 해결해보고 싶다면 Spring Security의 구조를 참고해봐도 좋을 것 같아요!

kkiseug reviewed (commented):
궁금한 점

kkiseug committed:
fix: 이름 길이 유효성 검사 로직 수정

- 에러 메시지를 '0-5글자'에서 '1-5글자'로 올바르게 수정

kkiseug committed:
test: Member 도메인 객체 테스트 구현

kkiseug committed:
refactor: 이름 길이 검증 로직 수정

- `name.length() <= 0`에서 `name.isEmpty()`로 수정

kkiseug committed:
refactor: createReservation 메서드에서 사용하지 않는 adminInfo 매개변수 제거

soominsohn commented on src/main/java/roomescape/auth/login/infrastructure/token/JwtTokenManager.java:
```diff
@@ -0,0 +1,57 @@
+    @Value("${secret.key}")
+    private String SECRET_KEY;
+
+    public String crateToken(final Long id, final String role) {
```
```suggestion
    public String createToken(final Long id, final String role) {
```
사소한 오타가 있네요~

kkiseug replied:
꼼꼼하지 못했네요.. 부끄럽습니다.

커밋: [f27821d](https://github.com/woowacourse/spring-roomescape-member/pull/281/commits/f27821d6ae7d9db7c5caf55156bb19974aa2b677)

soominsohn commented on src/main/java/roomescape/auth/login/infrastructure/token/JwtTokenManager.java:
```diff
@@ -0,0 +1,57 @@
+    private final static int EXPIRATION_TIME = 60 * 30 * 1000;
+
+    @Value("${secret.key}")
+    private String SECRET_KEY;
```
생성자주입 방식으로 사용하면 어떨까요~?

kkiseug replied:
바꾸기 전에는 기존의 방식을 생성자 주입으로 바꾸는 것이 어떤 이점을 줄지 몰랐는데, 직접 바꾸면서 적용해보니까 기존 방식의 필드는 final이 붙지않아 변경 가능성이 포함되어 있지만 생성자 방식을 사용함으로써 final 키워드를 붙여 변경 가능성을 차단할 수 있는 것이 장점이라고 생각했습니다! 앤지의 의도가 맞았을까요?~

커밋: [653cf98](https://github.com/woowacourse/spring-roomescape-member/pull/281/commits/653cf985b9881ef6a6399dea503e016666dad465)

soominsohn replied:
네 맞아요~ 👍 
생성자주입으로 만든김에 테스트코드도 만들어보면 어때요~?

soominsohn commented on src/main/java/roomescape/auth/login/infrastructure/token/JwtTokenManager.java:
```diff
@@ -0,0 +1,57 @@
+@Component
+public class JwtTokenManager {
+
+    private final static int EXPIRATION_TIME = 60 * 30 * 1000;
```
만료시간도 민감한정보기 때문에 설정파일에서 관리하고, 주입받으면 좋을 것 같습니다. 😄

kkiseug replied:
만료 시간도 민감한 정보에 포함되는군요! 몰랐어요. 바꾸겠습니당

커밋: [e4c7479](https://github.com/woowacourse/spring-roomescape-member/pull/281/commits/e4c747900147c3401b50e274132eeacedb8b9bce)

soominsohn commented on src/main/java/roomescape/auth/login/infrastructure/token/JwtTokenManager.java:
```diff
@@ -0,0 +1,57 @@
+                .build()
+                .parseSignedClaims(token)
+                .getPayload()
+                .getSubject());
```
JwtParser의 parseSignedClaims메서드는 아래 예외들을 throws하고 있는데요, 이 예외를 잡아 우리 프로그램에 적절한 예외로 변환해보는건 어떨까요~?
`Jws<Claims> parseSignedClaims(CharSequence jws) throws JwtException, IllegalArgumentException;`

kkiseug replied:
> Long.valueOf도 NumberFormatException을 throw하는데, 적절한 예외로 변환해주면 좋을 것 같아요.

해당 질문에 답변을 하면서 같이 해결하게 되었네요!! ☺️

soominsohn commented on src/main/java/roomescape/admin/infrastructure/JdbcAdminRepository.java:
```diff
@@ -0,0 +1,61 @@
+        SqlParameterSource param = new MapSqlParameterSource()
+                .addValue("id", id);
+
+        return namedParameterJdbcTemplate.queryForObject(sql, param, ROW_MAPPER);
+    }
```
queryForObject로 조회한 결과가 1건을 초과할 경우 예외가 발생하니, 이를 방지하거나 사전에 검증해주면 더 안전할 것 같아요.

kkiseug replied:
헉 맞습니다!! 사전에 검증하는 방식으로 해결했습니다! 어드민의 경우에는 회원가입 자체를 막고있어서 DB에 UNIQUE 키워드를 추가해주었고, 멤버의 경우에는 서비스에서 검증하고 동시에 DB에서도 UNIQUE 속성으로 막았습니다!

커밋: [86924f1](https://github.com/woowacourse/spring-roomescape-member/pull/281/commits/86924f1d3ebaf34a59034074a367cb83b485a37e)

soominsohn commented on src/main/java/roomescape/auth/login/infrastructure/token/JwtTokenManager.java:
```diff
@@ -0,0 +1,57 @@
+    public Long getId(final String token) {
+        validateValidToken(token);
+
+        return Long.valueOf(Jwts.parser()
```
Long.valueOf도 NumberFormatException을 throw하는데, 적절한 예외로 변환해주면 좋을 것 같아요.

kkiseug replied:
해당 예외 뿐만 아니라 사용하는 쪽에서 Token 관련 예외가 많이 발생할 수 있더라고요.

서명에 대한 예외 `SignatureException`
JWT 구조 자체의 문제 예외 `MalformedJwtException`
유효 기간이 지난 JWT에 대한 예외 `ExpiredJwtException` 등...

오히려 제가 Global에서 처리하고 있던 `SignatureException`는 이제 Deprecated되어 더 이상 사용되지 않는 것 같더라고요. 해당 부분들도 함께 수정했습니다!

커밋: [3e4ee4d](https://github.com/woowacourse/spring-roomescape-member/pull/281/commits/3e4ee4db499a6cce108d74b39525eacbe4704a39)
커밋: [9494b2c](https://github.com/woowacourse/spring-roomescape-member/pull/281/commits/9494b2c009a767de40727cc3e8a80ed5f1819d8b)

soominsohn replied:
맞아요~ 이번에 사용하신 JWT토큰 외에도, 다른 라이브러리를 사용하시게 된다면 어떤 예외를 반환하는지, 의도한 대로 동작을 할지 코드를 한번씩 까보면 좋습니다~

soominsohn commented on src/main/java/roomescape/admin/domain/Admin.java:
```diff
@@ -0,0 +1,36 @@
+public class Admin {
+
+    private final Long id;
+    private final String name;
+    private final String email;
+    private final String password;
```
Admin, Member로 나눈 구조가 흥미롭네요~
Member와 다르게 name, email, password는 객체가 아닌 원시값으로 관리하는 것도 의도된 설계일까요~?

kkiseug replied:
맞습니다! 제 설계에서 어드민의 경우에는 회원가입 자체를 막고 있어요! 제가 아니라 방탈출 사장님이 사용하는 계정이라고 생각했기 때문입니다. 그래서 서비스를 제공하는 쪽에서 DB에 직접 사장님 계정을 넣어주고 이를 제공하는 방식으로 생각하였습니다.

soominsohn commented on src/main/java/roomescape/common/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -1,35 +1,65 @@
+        return ResponseEntity.status(401).body(exceptionResponse);
+    }
+
+    @ExceptionHandler(UnauthorizedException.class)
+    public ResponseEntity<ExceptionResponse> handleUnauthorize(final UnauthorizedException exception, final HttpServletRequest request) {
+        ExceptionResponse exceptionResponse = new ExceptionResponse(
+                401, "[ERROR] " + exception.getMessage(), request.getRequestURI()
+        );
+
+        return ResponseEntity.status(401).body(exceptionResponse);
+    }
+
+    @ExceptionHandler(ForbiddenException.class)
+    public ResponseEntity<ExceptionResponse> handleForbidden(final ForbiddenException exception, final HttpServletRequest request) {
+        ExceptionResponse exceptionResponse = new ExceptionResponse(
+                403, "[ERROR] " + exception.getMessage(), request.getRequestURI()
+        );
+
+        return ResponseEntity.status(403).body(exceptionResponse);
+    }
```
HttpStatus 401과 403의 차이점은 무엇일까요?
또, 인증과 인가의 차이는 무엇일까요?

kkiseug replied:
401은 인증이고 403은 인가에 대한 에러 코드 입니다!

인증은 사용자에 대한 정보를 검증하는 것입니다. 요청으로 온 정보가 정말 어플리케이션에 등록된 사용자가 맞는지를 검증하는 것입니다. 예시로는 어떤 콘서트장에서 티켓과 함께 민증을 보여주는 경우가 해당될 수 있습니다!

인가는 인증된 사용자가 권한이 있는지를 나타내는 것입니다. 예를 들어 현재 제가 잠실 캠퍼스 13층에 있는 우아한 형제들에 권한이 없어 들어갈 수 없는 경우가 인가가 없는 것에 해당합니다!!

제 미션 구현에서는 어드민 권한이 있어야 들어갈 수 있는 페이지에 접속할 때 인가에 대한 예외를 던져줍니다!

soominsohn replied:
이해가 잘 되는 예시네요~ 👍

soominsohn commented on src/main/java/roomescape/common/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -1,35 +1,65 @@
+                401, "[ERROR] " + exception.getMessage(), request.getRequestURI()
+        );
+
+        return ResponseEntity.status(401).body(exceptionResponse);
```
```suggestion
        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(exceptionResponse);
```
http status code도 매직넘버보다 상수를 활용하면 좋을 것 같아요~

soominsohn commented on src/main/java/roomescape/auth/login/infrastructure/token/JwtTokenManager.java:
```diff
@@ -0,0 +1,57 @@
+    public String crateToken(final Long id, final String role) {
+        SecretKeySpec key = new SecretKeySpec(SECRET_KEY.getBytes(), "HmacSHA256");
+
+        return Jwts.builder()
```
Jwt의 구조는 어떻게 되고 각 필드의 역할은 무엇일까요~?

kkiseug replied:
JWT의 구조는 다음 3개로 구성됩니다.
- Header
- Payload
- Signature

`header.payload.signature`

Header에는 일반적으로 "alg", "typ" 필드가 들어가며 alg는 어떤 알고리즘을 사용해 서명하고 있는지를 나타내고, "typ"은 토큰의 종류에 대해서 나타냅니다.

Payload는 여러 클레임들을 포함합니다. 클레임은 사용자 및 추가적인 데이터에 대한 정보입니다. 이번에 조사하면서 알게된 것은 클레임도 public, private, registered 세 가지로 나뉜다는 점입니다!

registered는 기본적으로 정의된 클레임의 집합으로 각 필드는 다음과 같습니다.
- iss: JWT를 발행한 주체를 식별하기 위한 필드
- sub: JWT 토큰의 주제이며 해당 주제를 기반으로 클레임들이 구성된다고 이해했습니다. 그래서 현재는 멤버 id or 어드민 id로 구성하였습니다. 해당 id로 식별되는 엔티티들이기 때문에 id를 기반으로 해당 엔티티들의 역할을 지정해주는 것입니다!

처음에는 Admin과 Member를 이미 두 개의 객체로 나눠서 다루고 있는데 id를 Subject로 하는 것이 맞는가라는 의문이 들었습니다. 식별 자체는 id로 하는 것이 맞지만 서로의 id가 겹치기 때문에 추가적인 정보 Role을 통해 해주고 있기 때문입니다. 그래서 반대로 구성하는게 자연스럽지 않나 생각하기도 했습니다. 하지만, Subject는 전역적으로 고유해야 한다는 특징이 있기 때문에 식별값으로 하는 것이 더 의도에 맞다고 생각하였습니다. 앤지가 보기에는 어떤 구조가 자연스러운가요?

- exp: JWT가 수락되면 안되는 만료 날짜 또는 시간을 나타냅니다. 이 시간 이후로는 JWT 토큰이 verify되지 않습니다!
- aud: JWT가 의도한 수신자를 나타내는 값입니다.

public과 private는 각자 원하는 방식으로 이름을 정하고 값을 넣는 것은 동일하지만, 사용하는 범위가 다릅니다. private의 경우는 특정 어플리케이션 안에서 사용하는 것이고, public의 경우 공개적으로 jwt를 사용하는 경우 사용한다고 이해했어요! 정확한지는 모르겠지만, public claims는 깃허브 로그인, 네이버 로그인 같이 외부 로그인 연동에서 사용되는 것이 아닐까 싶습니다.

마지막으로 Signature는 header와 payload를 header에 적힌 알고리즘으로 서명한 것입니다! 이를 통해 서버에서 jwt 토큰에 대한 유효성을 검증할 수 있어요.

soominsohn replied:
id가 전역적으로 유일하다면 sub에 id를 넣어도 좋을 것 같아요~
sub에 들어가는 식별값과, db의 id는 일치하지 않아도 되기때문에, sub의 id에 `member, admin prefix` + `id`를 넣어주고 식별해도 좋을 것 같아요~. 하지만 role 클레임으로 보완하는 현재 방식도 충분히 괜찮은 것 같습니다.

JWT를 공부하고 잘 정리해주셨네요. 👍 
JWT말고 다른 인증방식에는 무엇이있고, 어떤 차이점이 있을지도 지금처럼 정리해두면 좋을 것 같아요!

soominsohn reviewed (changes requested):
안녕하세요 율무~
이번 미션도 정말 잘 해주셨네요! 
코멘트 몇 개 남겼으니 확인 부탁드려요~

kkiseug committed:
fix: 메서드 이름 오타 수정

- `crateToken` 메서드명을 `createToken`으로 수정

kkiseug committed:
refactor: HTTP 상태 코드를 HttpStatus 상수를 사용하도록 수정

kkiseug committed:
feat: 이메일 중복 검증 기능 추가

- 데이터베이스에서 email 컬럼에 UNIQUE 제약 조건 추가
- MemberService에 이메일 중복 여부 검증 로직 추가

kkiseug committed:
refactor: JwtTokenManager Value 프로퍼티 생성자 주입으로 변경

- 만료 시간도 properties에 함께 포함하도록 함

kkiseug committed:
refactor: JWT 처리 및 예외 처리 로직 개선

- 잘못된 형식, 만료, 기타 JWT 관련 예외에 대해 UnauthorizedException을 던지도록 수정
- 토큰의 ID 파싱 과정에서 숫자 형식 오류 발생 시 명확한 예외 처리 추가

kkiseug committed:
refactor: Jwt 예외 처리 로직 개선

- Jwt 관련 예외를 처리하기 위한 JwtExtractException 추가
- 중복된 예외 처리 로직을 extractClaims 메서드로 통합 및 재사용

kkiseug committed:
refactor: 토큰 관련 패키지 구조 개선

kkiseug reviewed (commented):
앤지!! 리뷰 감사합니다~!

kkiseug committed:
refactor: 인증 예외 처리 로직 분리

- UnauthorizedException과 ForbiddenException의 예외 처리를 AuthExceptionHandler로 분리

kkiseug committed:
refactor: 쓸모없는 메서드 제거

kkiseug committed:
feat: 비즈니스 예외 계층 추가

- 공통 예외 BusinessException 생성
- 이메일, 이름, 비밀번호 관련 예외 클래스 추가

kkiseug committed:
feat: 회원 도메인 검증 로직 변경

- Name, Email, Password에 대한 Null 및 빈 값 검증 로직 추가
- IllegalArgumentException을 커스텀 예외로 변경

kkiseug committed:
refactor: 값 객체 equals, hashcode 재정의

kkiseug committed:
fix: 이메일 중복 예외 클래스 변경

- IllegalArgumentException을 EmailException으로 변경

kkiseug committed:
refactor: 예약 생성 시 null 검증 로직 수정

- 예약 정보 검증을 위한 validateIsNonNull 메서드 추가
- BusinessException을 활용하여 null 값 처리

kkiseug committed:
refactor: 예약 예외 처리 로직 개선

- IllegalArgumentException을 ReservationException으로 변경

kkiseug committed:
feat: 예약 시간 관련 예외 처리 클래스 추가 및 검증 로직 개선

- ReservationTimeException 클래스 추가로 예외 관리
- 예약 시간 관련 검증 로직에서 IllegalArgumentException을 ReservationTimeException으로 변경
- null 값 검증 메서드 추가 및 BusinessException 활용

kkiseug committed:
refactor: Theme 관련 예외 처리 로직 개선

soominsohn reviewed (approved):
율무 안녕하세요~
이번 코멘트 반영도 잘 해주셨네요 👍 
학습한 내용도 남겨주셔서 고맙습니다~. 
다시 만나서 반가웠고, 율무 덕분에 많이 배웠습니다.
남은 미션들도 파이팅하세요! 응원하겠습니다~ 👋
